src/braket/ahs/analog_hamiltonian_simulation.py

```diff
@@ -63,6 +71,93 @@ def _hamiltonian_to_ir(self) -> ir.Hamiltonian:
+
+            Raises:
+                DiscretizeError: If unable to discretize the program.
+        """
+
+        # Gather resolution values
+
+        register_position_resolution = device.properties.paradigm.lattice.geometry.positionResolution
+        driving_parameters = device.properties.paradigm.rydberg.rydbergGlobal
+        driving_field_resolutions = {
+            'amplitude': {
+                'time': driving_parameters.timeResolution,
+                'value': driving_parameters.rabiFrequencyResolution
+            },
+            'detuning': {
+                'time': driving_parameters.timeResolution,
+                'value': driving_parameters.detuningResolution
+            },
+            'phase': {
+                'time': driving_parameters.timeResolution,
+                'value': driving_parameters.phaseResolution
+            }
+        }
+        shifting_parameters = device.properties.paradigm.rydberg.rydbergLocal
+        shifting_field_resolutions = {
+            'magnitude': {
+                'time': shifting_parameters.timeResolution,
+                'value': shifting_parameters.commonDetuningResolution,
+                'pattern': shifting_parameters.localDetuningResolution
+            }
+        }
+
+        # Discretize register
+
+        try:
+            discretized_register = self.register.discretize(register_position_resolution)
+        except:
+            raise DiscretizeError(f'Failed to discretize register {self.register}')
+
+        # Discretize Hamiltonian
+
+        discretized_hamiltonian = Hamiltonian()
+        for idx, term in enumerate(self.hamiltonian):
+            if isinstance(term, DrivingField):
+                resolutions = driving_field_resolutions
+                fields = dict(
+                    amplitude=term.amplitude,
+                    phase=term.phase,
+                    detuning=term.detuning
+                )
+            elif isinstance(term, ShiftingField):
+                resolutions = shifting_field_resolutions
+                fields = dict(
+                    magnitude=term.magnitude
+                )
+            else:
+                raise NotImplementedError(
+                    f'Only DrivingField and ShiftingField terms can be discretized.'
+                    f'Hamiltonian term {idx} is of type {type(term)}'
+                )
+
+            for field_name, original_field in fields.items():
+                try:
+                    time_res = resolutions[field_name]['time']
+                    value_res = resolutions[field_name]['value']
+                    pattern_res = resolutions[field_name].get('pattern')
+                    fields[field_name] = original_field.discretize(time_res, value_res, pattern_res)
+                except:
+                    DiscretizeError(f'Failed to discretize {field_name} of Hamiltonian term {idx}: {original_field}')
+
+            discretized_term = type(term)(**fields)
```

Comment:
This isn't guaranteed to work; it may be the case now, but future types may have constructors that take in different parameters than just their fields

Reply:
I don't know how to move forward here for making this more future proof. Do you have any suggestions?

Reply:
Why not just call the fields to discretize themselves, and pass in the device parameters?

Reply:
Not entirely sure how, but I would like to see how to simplify this part for discretizing all the fields.